### PR TITLE
Use jwts for OsfAuthHandler

### PR DESCRIPTION
**Pull request by chrisseto**
OSF-4733

### CODE REVIEW

File: waterbutler/auth/osf/handler.py [hunk 1]
```diff
@@ -36,7 +38,7 @@ def fetch(self, request, bundle):
         response = yield from aiohttp.request(
             'get',
             settings.API_URL,
-            params=bundle,
+            params={'payload': jwt.encode(bundle, settings.JWT_SECRET, algorithm=settings.JWT_ALGORITHM)},
```
Comment (icereval) [inline waterbutler/auth/osf/handler.py]: For added security we might as well set expiration on both sides.

https://github.com/jpadilla/pyjwt#expiration-time-claim

Comment (icereval) [inline waterbutler/auth/osf/handler.py]: Should use jwt.encode on the entire json object.

File: waterbutler/auth/osf/handler.py [hunk 2]
```diff
@@ -76,7 +78,7 @@ def get(self, resource, provider, request):
             response = yield from aiohttp.request(
                 'get',
                 settings.API_URL,
-                params=params,
+                params={'payload': jwt.encode(params, settings.JWT_SECRET, algorithm=settings.JWT_ALGORITHM)},
```
Comment (icereval) [inline waterbutler/auth/osf/handler.py]: Should use `jwt.encode` on the entire json object.

Comment (icereval) [inline waterbutler/auth/osf/handler.py]: For added security set expiration to 30s (https://github.com/jpadilla/pyjwt#expiration-time-claim)

File: waterbutler/auth/osf/handler.py [hunk 3]
```diff
@@ -48,7 +53,7 @@ def fetch(self, request, bundle):
                 data = yield from response.read()
             raise exceptions.AuthError(data, code=response.status)
 
-        return (yield from response.json())
+        return jwt.decode((yield from response.json()), settings.JWT_SECRET, algorithm=settings.JWT_ALGORITHM)
```
Comment (icereval) [inline waterbutler/auth/osf/handler.py]: Require exp - https://github.com/jpadilla/pyjwt#requiring-optional-claims

File: waterbutler/auth/osf/handler.py (hunk 1 above)
Comment (icereval) [inline waterbutler/auth/osf/handler.py]: This is very unlikely but I'll at least pose the question. Any concern for key collision, should we nest our data under data? https://github.com/jpadilla/pyjwt#support-of-registered-claim-names

**Comment by icereval** [comment]
First review done :pill:

File: waterbutler/auth/osf/settings.py [hunk 4]
```diff
@@ -5,4 +5,7 @@
 
 config = settings.get('OSF_AUTH_CONFIG', {})
 
+JWT_EXPIRATION = config.get('JWT_EXPIRATION', 15)
+JWT_ALGORITHM = config.get('JWT_ALGORITHM', 'HS256')
+JWT_SECRET = config.get('JWT_SECRET', 'ILiekTrianglesALot')
```
Comment (sloria) [inline waterbutler/auth/osf/settings.py]: Should we make this required? Maybe crash if it's not defined? Not sure how this is done elsewhere in WB.

**Comment by sloria** [comment]
Pass finished. Would be good to have some basic tests for this, but this could go in a separate PR.

File: waterbutler/auth/osf/settings.py (hunk 4 above)
Comment (icereval) [inline waterbutler/auth/osf/settings.py]: To simplify local development we've used to a non-crashing default value so user's do not need to have a configuration file.

Thought it certainly wouldn't hurt to issue a warning on important settings.

**Comment by chrisseto** [comment]
Merged via https://github.com/CenterForOpenScience/waterbutler/commit/71341de304a2054fa23708b12205e20a915a9044

**Pull request closed without merging**